contracts/stake-cw20-external-rewards/src/contract.rs

```diff
@@ -0,0 +1,2057 @@
+use crate::msg::{
+    ExecuteMsg, InfoResponse, InstantiateMsg, PendingRewardsResponse, QueryMsg, ReceiveMsg,
+};
+use crate::state::{
+    Config, RewardConfig, CONFIG, LAST_UPDATE_BLOCK, PENDING_REWARDS, REWARD_CONFIG,
+    REWARD_PER_TOKEN, USER_REWARD_PER_TOKEN,
+};
+use crate::ContractError;
+use crate::ContractError::{InvalidCw20, InvalidFunds, NoRewardsClaimable, Unauthorized};
+#[cfg(not(feature = "library"))]
+use cosmwasm_std::entry_point;
+
+use cosmwasm_std::{
+    from_binary, to_binary, Addr, BankMsg, Binary, Coin, CosmosMsg, Deps, DepsMut, Empty, Env,
+    MessageInfo, Response, StdError, StdResult, Uint128, Uint256, Uint512, WasmMsg,
+};
+use cw2::set_contract_version;
+use cw20::{Cw20ReceiveMsg, Denom};
+use stake_cw20::hooks::StakeChangedHookMsg;
+
+use cw20::Denom::Cw20;
+use std::cmp::min;
+use std::convert::TryInto;
+
+const CONTRACT_NAME: &str = "crates.io:stake-cw20-external-rewards";
+const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn instantiate(
+    deps: DepsMut,
+    _env: Env,
+    _info: MessageInfo,
+    msg: InstantiateMsg,
+) -> Result<Response<Empty>, ContractError> {
+    set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+
+    let owner = msg.owner.map(|a| deps.api.addr_validate(&a)).transpose()?;
+    let manager = msg
+        .manager
+        .map(|a| deps.api.addr_validate(&a))
+        .transpose()?;
+
+    let reward_token = match msg.reward_token {
+        Denom::Native(denom) => Denom::Native(denom),
+        Cw20(addr) => Cw20(deps.api.addr_validate(&addr.to_string())?),
+    };
+
+    let config = Config {
+        owner,
+        manager,
+        staking_contract: deps.api.addr_validate(&msg.staking_contract)?,
```

Comment:
Could we run a `GetConfig {}` query here to smoke check that the address supplied is actually a staking contract?

Reply:
good call out, used total staked query as it is more specific to the staking contract.